services/user/Tokens/service/src/tables.rs

```diff
@@ -0,0 +1,505 @@
+
+            Balance::get_or_new(receiver, self.id).add_balance(amount);
+        }
+
+        pub fn burn(&mut self, amount: Quantity, burnee: AccountNumber) {
+            check(amount.value > 0, "burn quantity must be greater than 0");
+            self.current_supply = self.current_supply - amount;
+            self.save();
+
+            Balance::get_or_new(burnee, self.id).sub_balance(amount);
+        }
+    }
+
+    #[ComplexObject]
+    impl Token {
+        pub async fn owner(&self) -> AccountNumber {
+            self.nft_holder()
+        }
+
+        pub async fn precision(&self) -> Precision {
+            self.precision.try_into().unwrap()
+        }
+
+        pub async fn current_supply(&self) -> Decimal {
+            Decimal::new(self.current_supply, self.precision.try_into().unwrap())
+        }
+
+        pub async fn max_supply(&self) -> Decimal {
+            Decimal::new(self.max_supply, self.precision.try_into().unwrap())
+        }
+
+        pub async fn is_unburnable(&self) -> bool {
+            self.settings().is_unburnable()
+        }
+
+        pub async fn is_unrecallable(&self) -> bool {
+            self.settings().is_unrecallable()
+        }
+
+        pub async fn is_untransferable(&self) -> bool {
+            self.settings().is_untransferable()
+        }
+    }
+
+    #[table(name = "BalanceTable", index = 2)]
+    #[derive(Serialize, Deserialize, ToSchema, Fracpack, Debug, SimpleObject)]
+    #[graphql(complex)]
+    pub struct Balance {
+        pub account: AccountNumber,
+        pub token_id: TID,
+        #[graphql(skip)]
+        pub balance: Quantity,
+    }
+
+    impl Balance {
+        #[primary_key]
+        fn pk(&self) -> (AccountNumber, TID) {
+            (self.account, self.token_id)
+        }
+
+        fn new(account: AccountNumber, token_id: TID) -> Self {
+            Token::get_assert(token_id);
+            Self {
+                account,
+                token_id,
+                balance: 0.into(),
+            }
+        }
+
+        pub fn get(account: AccountNumber, token_id: TID) -> Option<Self> {
+            BalanceTable::new().get_index_pk().get(&(account, token_id))
+        }
+
+        pub fn add(account: AccountNumber, token_id: TID) -> Self {
+            check_none(Self::get(account, token_id), "balance already exists");
+            let mut instance = Self::new(account, token_id);
+            instance.save();
+            instance
+        }
+
+        pub fn get_or_new(account: AccountNumber, token_id: TID) -> Self {
+            Self::get(account, token_id).unwrap_or(Self::new(account, token_id))
+        }
+
+        fn save(&mut self) {
+            let table = BalanceTable::new();
+            table.put(&self).unwrap();
+        }
+
+        fn add_balance(&mut self, quantity: Quantity) {
+            self.balance = self.balance + quantity;
+            self.save();
+        }
+
+        fn sub_balance(&mut self, quantity: Quantity) {
```

Comment:
This is missing a check on `Holder::is_keep_zero_balances` and determining whether to delete the balance record or not.

Reply:
I only ever thought `keep_zero_balances` was to be applied to SharedBalances, this is to spare Alice of the RAM expense of frequent sending to a bunch of users who drop the record by debiting each time forcing Alice to keep paying for it again and again, however, for her own balance this seems less significant. 

Also, you said you want the `TokenHolder` table removed entirely and the flags moved to `Balance`, but if we're by default dropping your `Balance` unless you've manually set `keep_zero_balances`  to true, you can accidentally delete your own `auto_debit` setting for a token just by sending your entire balance elsewhere. Seems funky.